### PR TITLE
refactor(pageserver): rely on serde derive for toml deserialization

### LINKED ISSUE #3682: pageserver: Difficulty of adding config options
We've been discussing that adding new fields to TenantConf or PageserverConf is quite multiline, and is repeated in neon_local. This has lead to it being too cumbersome to use for example feature flags.

Adding of a new structured tenantconf option difficulties and tradeoffs can be seen on [d14473b](https://github.com/neondatabase/neon/pull/3552/commits/d14473bdeb3e863c1dc2d4da72ca34b582ef8a88) (sadly squashed).

Related: https://github.com/neondatabase/neon/pull/3597#discussion_r1114347866

### PR DESCRIPTION
**Pull request by problame**
This PR simplifies the pageserver configuration parsing as follows:

* introduce the `pageserver_api::config::ConfigToml` type
* implement `Default` for `ConfigToml`
* use serde derive to do the brain-dead leg-work of processing the toml document
  * use `serde(default)` to fill in default values
* in `pageserver` crate:
  * use `toml_edit` to deserialize the pageserver.toml string into a `ConfigToml`
  * `PageServerConfig::parse_and_validate` then
    * consumes the `ConfigToml`
    * destructures it exhaustively into its constituent fields
    * constructs the `PageServerConfig`

The rules are:

* in `ConfigToml`, use `deny_unknown_fields` everywhere
* static default values go in `pageserver_api`
* if there cannot be a static default value (e.g. which default IO engine to use, because it depends on the runtime), make the field in `ConfigToml` an `Option`
* if runtime-augmentation of a value is needed, do that in `parse_and_validate`
  * a good example is `virtual_file_io_engine` or `l0_flush`, both of which need to execute code to determine the effective value in `PageServerConf`

The benefits:

* massive amount of brain-dead repetitive code can be deleted
* "unused variable" compile-time errors when removing a config value, due to the exhaustive destructuring in `parse_and_validate`
* compile-time errors guide you when adding a new config field

Drawbacks:

* serde derive is sometimes a bit too magical
* `deny_unknown_fields` is easy to miss

Future Work / Benefits:
* make `neon_local` use `pageserver_api` to construct `ConfigToml` and write it to `pageserver.toml`
  * This provides more type safety / coompile-time errors than the current approach.

### Refs

Fixes #3682 

### Future Work

* `remote_storage` deser doesn't reject unknown fields https://github.com/neondatabase/neon/issues/8915
* clean up `libs/pageserver_api/src/config.rs` further
  * break up into multiple files, at least for tenant config
    * move `models` as appropriate / refine distinction between config and API models / be explicit about when it's the same
  * use `pub(crate)` visibility on `mod defaults` to detect stale values

### CODE REVIEW

File: libs/pageserver_api/src/config.rs [hunk 1]
```diff
@@ -29,3 +36,497 @@ pub struct NodeMetadata {
     #[serde(flatten)]
     pub other: HashMap<String, serde_json::Value>,
 }
+
+#[serde_as]
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
+#[serde(default)]
+pub struct ConfigToml {
```
Comment (problame) [inline libs/pageserver_api/src/config.rs:43]: missing `deny_unnknown_fields`

File: pageserver/src/config.rs [hunk 2]
```diff
@@ -1019,7 +427,14 @@ impl PageServerConf {
             );
         }
 
-        conf.default_tenant_conf = t_conf.merge(TenantConf::default());
+        IndexEntry::validate_checkpoint_distance(conf.default_tenant_conf.checkpoint_distance)
+            .map_err(|msg| anyhow::anyhow!("{msg}"))
```
Comment (koivunej) [inline pageserver/src/config.rs:431]: I am a bit unsure what does this do, but it looks weird. If we really have a `Result<(), &'static str>` returning fn, then `anyhow::Error::msg`
```suggestion
            .map_err(anyhow::Error::msg)
```

File: pageserver/src/bin/pageserver.rs [hunk 3]
```diff
@@ -223,13 +223,13 @@ fn initialize_config(
...
 
-    let config: toml_edit::Document = match std::fs::File::open(cfg_file_path) {
+    let config: pageserver_api::config::ConfigToml = match std::fs::File::open(cfg_file_path) {
         Ok(mut f) => {
             let md = f.metadata().context("stat config file")?;
             if md.is_file() {
                 let mut s = String::new();
                 f.read_to_string(&mut s).context("read config file")?;
-                s.parse().context("parse config file toml")?
+                toml_edit::de::from_str(&s).context("deserialize config file toml")?
```
Comment (koivunej) [inline pageserver/src/bin/pageserver.rs:232]: I think here we could use the the serde-whatever-track-depth (I'll look this up soon).

File: pageserver/src/bin/pageserver.rs [hunk 4]
```diff
@@ -223,13 +223,13 @@ fn initialize_config(
...
             } else {
                 anyhow::bail!("directory entry exists but is not a file: {cfg_file_path}");
             }
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/bin/pageserver.rs:235]: It seems we are doing the same with identity, unsure why.

**Review by koivunej — APPROVED** [review]
Very nice, thanks.

**Review by arpad-m — APPROVED** [review]

File: pageserver/src/config.rs (hunk 2 above)
Comment (problame) [reply pageserver/src/config.rs]: b598782b430b4f301a9305b577e0771ca7f811fb

File: pageserver/src/bin/pageserver.rs (hunk 3 above)
Comment (koivunej) [reply pageserver/src/bin/pageserver.rs]: At least the error now looks good:

```
Error: deserialize config toml

Caused by:
    tenant_config.gc_period: TOML parse error at line 11, column 13
       |
    11 | gc_period = "0"
       |             ^^^
    invalid value: string "0", expected a duration


Stack backtrace:
   0: pageserver::initialize_config
             at /home/joonas/src/neon/work/pageserver/src/bin/pageserver.rs:229:23
      pageserver::main
             at /home/joonas/src/neon/work/pageserver/src/bin/pageserver.rs:95:16
```

Without we get:

```
Error: deserialize config toml

Caused by:
    TOML parse error at line 11, column 13
       |
    11 | gc_period = "0"
       |             ^^^
    invalid value: string "0", expected a duration
```

which is strictly worse.

So I guess I need to check the other uses next...

Comment (koivunej) [reply pageserver/src/bin/pageserver.rs]: Oki, so the ` impl TryFrom<toml_edit::Item> for TenantConfOpt` in pageserver/src/tenant/config.rs:444 is probably/hopefully dead code now.

I think it's how the deserialization takes a different route, or maybe it's working around some annoyance which has gotten fixed in toml_edit. I remember adding that for a good reason...

**Review by koivunej — APPROVED** [review]

File: pageserver/src/bin/pageserver.rs (hunk 3 above)
Comment (problame) [reply pageserver/src/bin/pageserver.rs]: https://github.com/neondatabase/neon/issues/8918

File: test_runner/regress/test_timeline_size.py [hunk 5]
```diff
@@ -1062,7 +1062,7 @@ def eager_tenant_is_active():
 @pytest.mark.parametrize("activation_method", ["endpoint", "branch", "delete"])
 def test_lazy_attach_activation(neon_env_builder: NeonEnvBuilder, activation_method: str):
     # env.initial_tenant will take up this permit when attaching with lazy because of a failpoint activated after restart
-    neon_env_builder.pageserver_config_override = "concurrent_tenant_warmup = '1'"
+    neon_env_builder.pageserver_config_override = "concurrent_tenant_warmup = 1"
```
Comment (koivunej) [inline test_runner/regress/test_timeline_size.py:1065]: we could post to the oss channel in discord about this.

**Review by koivunej — APPROVED** [review]

File: test_runner/regress/test_timeline_size.py (hunk 5 above)
Comment (problame) [reply test_runner/regress/test_timeline_size.py:1065]: Will do

Comment (problame) [reply test_runner/regress/test_timeline_size.py:1065]: https://discord.com/channels/1176467419317940276/1194680868439076874/1281238083987177532

**Pull request merged**